src/libostree/ostree-async-progress.c

```diff
@@ -424,6 +424,42 @@ ostree_async_progress_set_uint64 (OstreeAsyncProgress       *self,
   ostree_async_progress_set_variant (self, key, g_variant_new_uint64 (value));
 }
 
+/**
+ * ostree_async_progress_copy_state:
+ * @self: An #OstreeAsyncProgress to copy from
+ * @dest: An #OstreeAsyncProgress to copy to
+ *
+ * Atomically copies all the state from @self to @dest, without invoking the
+ * callback.
+ * This is used for proxying progress objects across different #GMainContexts.
+ *
+ * Since: 2019.6
+ */
+void
+ostree_async_progress_copy_state (OstreeAsyncProgress *self,
+                                  OstreeAsyncProgress *dest)
+{
+  g_return_if_fail (OSTREE_IS_ASYNC_PROGRESS (self));
+  g_return_if_fail (OSTREE_IS_ASYNC_PROGRESS (dest));
+
+  g_mutex_lock (&self->lock);
+
+  if (self->dead)
+    goto out;
+
+  GLNX_HASH_TABLE_FOREACH_KV (self->values, void *, key, GVariant *, value)
+    {
+      if (value)
```

Comment:
Is this on purpose? Seems like if the value for a key is `NULL` in the original table, we'd want to copy that state too, no? Maybe just:

```c
if (value)
  g_variant_ref (value);
g_hash_table_replace (dest->values, key, value); 
```

?

Reply:
This was on purpose since `ostree_async_progress_get()` will crash if the value for a requested key is `NULL`, but I don't think it makes much difference either way. I guess if you had an invalid progress object before, it may as well be invalid after the copy, otherwise it could mask bugs, so maybe this is a bit better. I'll test the flatpak fix with this change, and if no surprises, update this PR.

Reply:
OK, updated.